expression/builtin_compare.go

```diff
@@ -294,11 +296,83 @@ func isTemporalColumn(expr Expression) bool {
 	return true
 }
 
+func tryToConvertConstantInt(con *Constant, ctx context.Context) *Constant {
+	if con.GetTypeClass() == types.ClassInt {
+		return con
+	}
+	sc := ctx.GetSessionVars().StmtCtx
+	i64, err := con.Value.ToInt64(sc)
+	if err != nil {
+		return con
+	}
+	return &Constant{
+		Value:   types.NewIntDatum(i64),
+		RetType: types.NewFieldType(mysql.TypeLonglong),
+	}
+}
+
+// refineConstantArg changes the constant argument to it's ceiling or flooring result by the given op.
+func refineConstantArg(con *Constant, op opcode.Op, ctx context.Context) *Constant {
+	sc := ctx.GetSessionVars().StmtCtx
+	i64, err := con.Value.ToInt64(sc)
+	if err != nil {
+		log.Warnf("failed to refine this expression, the constant argument is %s", con)
+		return con
+	}
+	datumInt := types.NewIntDatum(i64)
```

Comment:
``` sql
create table t(a bigint unsigned);
select a < cast(-1 as unsigned);
```

`a < cast(-1 as unsigned)` will be refined as `a < -1`?
`a < cast(-1 as unsigned)` will be compared as `unsigned` int,
but `a < -1` will be compared as `signed` int.